src/network/core/http_winhttp.cpp

```diff
@@ -258,14 +264,17 @@ void NetworkHTTPRequest::Connect()
  */
 bool NetworkHTTPRequest::Receive()
 {
-	if (this->callback->IsCancelled()) {
+	if (this->callback.cancelled && !this->finished) {
 		Debug(net, 1, "HTTP request failed: cancelled by user");
 		this->finished = true;
-		this->callback->OnFailure();
-		return true;
+		this->callback.OnFailure();
+		/* Fall-through, as we are waiting for IsQueueEmpty() to happen. */
 	}
 
-	return this->finished;
+	/* Only return true if the queue was also dequeued. */
+	if (!this->finished) return false;
+	if (!this->callback.IsQueueEmpty()) return false;
+	return true;
```

Comment:
I think the following reads easier:
```suggestion
	return this->finished && this->callback.IsQueueEmpty();
```